### PR TITLE
Feat: Adapter Docs

### PR DESCRIPTION
**Pull request by ElusAegis**
Written docs for the adapters: both Circom and Halo2, as well as provide updated README and added benchmark results for Halo2.

As there are plans to integrate multiple adapters, created a different section for adapters in the docs.

### CODE REVIEW

**Comment by ElusAegis** [comment]
I feel like right now the IOS build error is happening way more often

**Review by vivianjeng — CHANGES_REQUESTED** [review]

File: docs/docs/adapters/overview.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+```toml
+[features]
+default = ["mopro-ffi/circom", "mopro-ffi/halo"]
+```
+
+For adapter specific dependencies, please refer to each adapter's documentation.
+
+## Supported Adapters
+
+- [Circom](/adapters/circom.md) - ["mopro-ffi/circom"]
```
Comment (vivianjeng) [inline docs/docs/adapters/overview.md:38]:
```suggestion
- [Circom](/adapters/circom.md) - `["mopro-ffi/circom"]`
```
It looks better with code block

File: docs/docs/performance.md [hunk 2]
```diff
@@ -4,115 +4,122 @@ sidebar_position: 5
...
+|       Average        |                     2908.6 ms                     |                         2324.4 ms                         |                 10304.8 ms                  |
+|        Stdev         |                     112.9 ms                      |                          67.1 ms                          |                  605.5 ms                   |
+| Comparing to snarkjs |                       ~3.5x                       |           <font color="FFB546">**~4.4x**</font>           |                      -                      |
 
 <iframe width="484" height="266" seamless frameborder="0" scrolling="no" src="https://docs.google.com/spreadsheets/d/e/2PACX-1vTbLHpEzT1ybhta5NVQrDQuOLwgGzLGpm2amiWgLRu0l9ZvXMNtNZ-DWIlySL0zO30UWn_nZvkfaQWY/pubchart?oid=1171109874&amp;format=interactive"></iframe>
 </details>
 
-**Details:** [Spreadsheet of Circom benchmark (iOS)](https://docs.google.com/spreadsheets/d/1MFABmsYSUsWDmhbjleqhBXk7nkYwhu589yK-CHtRkNI/edit?usp=sharing)
+**Details:
+** [Spreadsheet of Circom benchmark (iOS)](https://docs.google.com/spreadsheets/d/1MFABmsYSUsWDmhbjleqhBXk7nkYwhu589yK-CHtRkNI/edit?usp=sharing)
```
Comment (vivianjeng) [inline docs/docs/performance.md:83]:
```suggestion
**Details:** [Spreadsheet of Circom benchmark (iOS)](https://docs.google.com/spreadsheets/d/1MFABmsYSUsWDmhbjleqhBXk7nkYwhu589yK-CHtRkNI/edit?usp=sharing)
```

it shouldn't be a new line or new space after `Details:`

File: README.md [hunk 3]
```diff
@@ -2,17 +2,23 @@
...
+## Using a pre-built app library
 
 There are pre-built examples projects for iOS and Android [here](https://github.com/vimwitch/mopro-app).
 
-## Building an app library
+## Building your own app library
 
-Mopro works by providing a static library and an interface for your app to build proofs. Before you start this tutorial you should have a zkey and wasm file generated by circom.
+Mopro works by providing a static library and an interface for your app to build proofs. Before you start this tutorial
+you should have a zkey and wasm file generated by circom.
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply README.md:21]: No, my IDEA is strict formating about forming lines of a certain length, forcefully doing a new line on them. This should still be rendered as one line in MD.

If this is confusing, I can merge it into one line.

File: docs/docs/adapters/circom.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+that builds this library with Circom proofs.
+
+In your `Cargo.toml` file, ensure the `mopro-ffi/circom` feature is activated:
+
+```toml
+[features]
+default = ["mopro-ffi/circom"]
+```
+
+## Witness Generation Functions
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply docs/docs/adapters/circom.md:24]: No worries 😉

File: docs/docs/adapters/circom.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    // ...
+}
+```
+
+And then you can automatically generate the witness functions for all the circuits in the specified folder.
+
+To do so, in the `lib.rs` file, you can add the following:
+
+```rust
+rust_witness::generate_witnesses!("your_circuit_name");
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply docs/docs/adapters/circom.md:72]: Good catch, thank you!

File: docs/docs/adapters/halo2.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+Explore how the Halo2 adapter is implemented by checking out this sample
+project: [Halo2 RSA Mopro on GitHub](https://github.com/ElusAegis/halo2-rsa-mopro).
+
+## Setting Up the Rust Project
+
+Start by following the general instructions in the [Rust Setup Guide](/getting-started/rust-setup.md) to create a new
+Rust project for building libraries with Circom proofs. However, note these specific adjustments for Halo2:
+
+In your `Cargo.toml` file, ensure the `mopro-ffi/halo` feature is activated:
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply docs/docs/adapters/halo2.md:17]: Well, yes, but to enable it you need to specify ut on the `mopro-ffi`. I will make it more explicit

File: docs/docs/adapters/halo2.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+```toml
+[features]
+default = ["mopro-ffi/halo"]
+```
+
+You can also remove the rust-witness dependency as it is unnecessary for Halo2 circuits. Likewise, remove the
+`rust_witness::generate_witnesses!("...")` macro from the `build.rs` file.
```
[the remark this answers is not in the document]
Comment (ElusAegis) [reply docs/docs/adapters/halo2.md:25]: Correct, I have also fixed the circom documentation where there was a mistake. I should take more care with Copilot

File: docs/docs/performance.md (hunk 2 above)
Comment (ElusAegis) [reply docs/docs/performance.md:83]: Sorry for changes to the file - again this was due to my IDEA formating action on save. Should I rever all the chagnes to the tables, or only fix this one?

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (vivianjeng) [reply README.md:21]: I always use **Word Wrap** in vscode
(so it doesn't matter if we use different size of screen)
I think this line is not confusing but the following lines are 😅
I prefer to keep them in one line 🙏🏻 thank you

File: docs/docs/adapters/circom.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
+# Circom Adapter
+
+Mopro supports the integration of Circom circuits. For this, you need to have pre-built `zkey` and `wasm` files for your
+circuits. You can find more information on how to generate these files in
+the [Circom documentation](https://docs.circom.io).
+
+## Sample Project
+
+Explore how the Circom adapter is implemented by checking out this sample
+project: [mopro-app](https://github.com/vimwitch/mopro-app).
```
Comment (vivianjeng) [inline docs/docs/adapters/circom.md:10]: can we use https://github.com/zkmopro/mopro/tree/main/test-e2e
this is where we maintain (and test) each adapter

File: docs/docs/adapters/halo2.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
+# Halo2 Adapter
+
+Mopro supports the integration of Halo2 circuits, allowing for both the Halo2 library from Zcash and the PSE's Halo2
+fork. To effectively work with Halo2 circuits, you need to understand how to generate proving and verifying keys as well
+as how Halo2 circuits work. For more details, refer to the [Halo2 documentation](https://zcash.github.io/halo2/).
+
+## Sample Project
+
+Explore how the Halo2 adapter is implemented by checking out this sample
+project: [Halo2 RSA Mopro on GitHub](https://github.com/ElusAegis/halo2-rsa-mopro).
```
Comment (vivianjeng) [inline docs/docs/adapters/halo2.md:10]: when you complete `halo2-app` you can also update the link :)

File: docs/docs/performance.md [hunk 10]
```diff
@@ -128,21 +134,49 @@ Benchmarks on an Macbook Pro M1 Max (2021).
...
+|        Stdev         |                       18.7 ms                       |                      21.3 ms                      |                         266.1 ms                          |                   69.3 ms                   |
+| Comparing to snarkjs |        <font color="FFB546">**~2.3**</font>         |                       ~2.1x                       |                           ~2.1x                           |                      -                      |
 
 <iframe width="472" height="279" seamless frameborder="0" scrolling="no" src="https://docs.google.com/spreadsheets/d/e/2PACX-1vQOuS5abdzvh6znXORvSo7M85ubmDpSmE3C1Zs_l56wd25lMK4FZPEWOaCp7WrOlIjc3jEcWa0lfiy9/pubchart?oid=1443184132&amp;format=interactive"></iframe>
 </details>
 
-**Details:** [Spreadsheet of Circom benchmark (macOS)](https://docs.google.com/spreadsheets/d/1irKg_TOP-yXms8igwCN_3OjVrtFe5gTHkuF0RbrVuho/edit?usp=sharing)
+**Details:
+** [Spreadsheet of Circom benchmark (macOS)](https://docs.google.com/spreadsheets/d/1irKg_TOP-yXms8igwCN_3OjVrtFe5gTHkuF0RbrVuho/edit?usp=sharing)
```
Comment (vivianjeng) [inline docs/docs/performance.md:155]: should not be a new line

File: docs/docs/adapters/circom.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+# ...
+rust-witness = { git = "https://github.com/vimwitch/rust-witness.git" }
+```
+
+#### Configuring the path to the `.wasm` circuit files in the `build.rs`
+
+Then you need to add the `generate_witnesses` to the `build.rs` file and specify the path to your circom circuits to
+compile them.
+
+The value passed to the `generate_witnesses` macro should be the path to the folder containing the `.wasm` files for
```
Comment (vivianjeng) [inline docs/docs/adapters/circom.md:62]: here is wrong too

File: docs/docs/adapters/circom.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+should be called in the `lib.rs` file of your project, after the `mopro_ffi::app()` macro, and it should contain a list
+of tuples, where the first element is the name of the `zkey` file and the second element is the witness generation
+function.
+
+For example:
+
+```rust
+mopro_ffi::set_circom_circuits! {
+    ("your_circuit.zkey", your_circuit_wtns_gen_fn),
+    ("another_circuit.zkey", another_circuit_wtns_gen_fn),
```
Comment (vivianjeng) [inline docs/docs/adapters/circom.md:120]: can also use the example of `multiplier2`?

File: docs/docs/adapters/overview.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+```toml
+[features]
+default = ["mopro-ffi/circom"]
+```
+
+To mix different adapters, you can enable multiple features:
+
+```toml
+[features]
+default = ["mopro-ffi/circom", "mopro-ffi/halo"]
```
Comment (vivianjeng) [inline docs/docs/adapters/overview.md:31]: haven't been solved here?

File: docs/docs/adapters/overview.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+```
+
+For adapter specific dependencies, please refer to each adapter's documentation.
+
+## Supported Adapters
+
+- [Circom](/adapters/circom.md) - `["mopro-ffi/circom"]`
+- [Halo](/adapters/halo2.md) - `["mopro-ffi/halo"]`
```
Comment (vivianjeng) [inline docs/docs/adapters/overview.md:39]: same here

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ElusAegis) [reply README.md:21]: I use RustResolver by Jetbrains .. they have a hard wrap at 120 symbols by default. I have turned it off to fix it

File: docs/docs/adapters/overview.md (hunk 13 above)
Comment (ElusAegis) [reply docs/docs/adapters/overview.md:31]: Did a regex search, that is the last one

File: docs/docs/adapters/circom.md (hunk 8 above)
Comment (ElusAegis) [reply docs/docs/adapters/circom.md:10]: Do we then need a sample Halo2 app? As we have in this sample both Halo2 and Circom adapter?

File: docs/docs/adapters/overview.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+[features]
+default = ["mopro-ffi/circom", "mopro-ffi/halo2"]
+```
+
+For adapter specific dependencies, please refer to each adapter's documentation.
+
+## Supported Adapters
+
+- [Circom](/adapters/circom.md) - `["mopro-ffi/circom"]`
+- [Halo](/adapters/halo2.md) - `["mopro-ffi/halo2"]`
```
Comment (vivianjeng) [inline docs/docs/adapters/overview.md:36]:
```suggestion
- [Halo2](/adapters/halo2.md) - `["mopro-ffi/halo2"]`
```

File: docs/docs/performance.md (hunk 10 above)
Comment (vivianjeng) [inline docs/docs/performance.md:155]:
```suggestion
**Details:** [Spreadsheet of Circom benchmark (macOS)](https://docs.google.com/spreadsheets/d/1irKg_TOP-yXms8igwCN_3OjVrtFe5gTHkuF0RbrVuho/edit?usp=sharing)
```

Comment (ElusAegis) [reply docs/docs/performance.md:155]: I sware I fixed it. Sorry for the autocorrection doing it again!

**Pull request merged**